src/vos/vos_io.c

```diff
@@ -2106,33 +2107,34 @@ vos_update_end(daos_handle_t ioh, uint32_t pm_ver, daos_key_t *dkey, int err,
 			err = -DER_TX_RESTART;
 		}
 	}
+
+	if (err == 0)
+		vos_ts_set_upgrade(ioc->ic_ts_set);
+
+	if (err == -DER_NONEXIST || err == -DER_EXIST || err == 0) {
```

Comment:
The reason this was after the vos_tx_end is because vos_tx_end can fail even if we've succeeded to this point.   I guess it's not a huge issue but we will be updating timestamps for a transaction that doesn't commit.  Maybe move it after vos_tx_end if it needs to move before your new stuff ?

Reply:
Why need to update the timestamp if the transaction is failed?

On the other hand, I hope we can keep consistent logic for both update and punch case.